editor.planx.uk/src/pages/FlowEditor/components/Sidebar/EditHistory/Timeline.tsx

```diff
@@ -0,0 +1,171 @@
+import RestoreOutlined from "@mui/icons-material/RestoreOutlined";
+import Timeline from "@mui/lab/Timeline";
+import TimelineConnector from "@mui/lab/TimelineConnector";
+import TimelineContent from "@mui/lab/TimelineContent";
+import TimelineDot from "@mui/lab/TimelineDot";
+import TimelineItem, { timelineItemClasses } from "@mui/lab/TimelineItem";
+import TimelineSeparator from "@mui/lab/TimelineSeparator";
+import Box from "@mui/material/Box";
+import IconButton from "@mui/material/IconButton";
+import Tooltip from "@mui/material/Tooltip";
+import Typography from "@mui/material/Typography";
+import { OT } from "@planx/graph/types";
+import { useStore } from "pages/FlowEditor/lib/store";
+import { formatLastEditDate } from "pages/FlowEditor/utils";
+import React, { useState } from "react";
+import { FONT_WEIGHT_SEMI_BOLD } from "theme";
+import { CommentHistoryItem, HistoryItem, OperationHistoryItem, PublishHistoryItem } from ".";
+import { CommentTimelineItem, OperationTimelineItem, PublishTimelineItem } from "./TimelineItems";
+
+interface EditHistoryTimelineProps {
+  events: HistoryItem[];
+}
+
+export const EditHistoryTimeline = ({ events = [] }: EditHistoryTimelineProps) => {
+  const [focusedOpIndex, setFocusedOpIndex] = useState<number | undefined>(
+    undefined,
+  );
+
+  const [flow, teamSlug, canUserEditTeam, undoOperation, deleteFlowComment] = useStore(
+    (state) => [
+      state.flow,
+      state.teamSlug,
+      state.canUserEditTeam,
+      state.undoOperation,
+      state.deleteFlowComment,
+    ],
+  );
+
+  const handleUndo = (i: number) => {
+    // Get all events since & including the selected one
+    const eventsToUndo = events.slice(0, i + 1);
+
+    const operationsToUndo = eventsToUndo.filter((event) => event.type === "operation") as OperationHistoryItem[];
+    const commentsToDelete = eventsToUndo.filter((event) => event.type === "comment") as CommentHistoryItem[];
+
+    // Make a flattened list of operations, with the latest operations first
+    const operationsData: Array<OT.Op[]> = [];
+    operationsToUndo.map((op) => operationsData.unshift(op.data));
+    const flattenedOperationsData: OT.Op[] = operationsData?.flat(1);
+
+    // Then undo all operations by applying a single new inverse operation on top
+    undoOperation(flattenedOperationsData);
+
+    // Also delete each comment in the undo scope
+    commentsToDelete.forEach((comment) => {
+      deleteFlowComment(comment.id);
```

Comment:
This might well be correct and we don't need to `await` this, but just flagging it's an `async` operation within a `forEach` so this won't get handled consistently.

Reply:
Async/await now added, my oversight - thanks for spotting!